### PR TITLE
Feature/65 Change package to become a plugin and add custom rules

### LINKED ISSUE #65: Create custom eslint rule
@adamgajzlerowicz commented on [Tue Mar 03 2020](https://github.com/pmedianetwork/adverity-presense-frontend/issues/327)

PR https://github.com/pmedianetwork/eslint-config/pull/64

---

**About**
We require an implementation of a custom eslint rule. 
This rule will almost a fork of [no-relative-parent-imports](https://github.com/benmosher/eslint-plugin-import/blob/master/src/rules/no-relative-parent-imports.js), with a difference that it will allow importing from one level up.

---

**Tasks**

- [x] lint rule is implemented in our own repository
- [x] rule is added do eslint repository as a warning for now
- [ ] ~warnings are fixed within presence repository~

---

@adamgajzlerowicz commented on [Wed Mar 04 2020](https://github.com/pmedianetwork/adverity-presense-frontend/issues/327#issuecomment-594459367)

Here is the plugin code:
https://github.com/adamgajzlerowicz/sensible-imports

For a prototype I used my own repo, but once we decide we are going to use it I will move it to company namespace.

Here is plugin test result
image

It can be used like this:
https://github.com/pmedianetwork/eslint-config/blob/sensible-imports/index.js#L31

Issues are reported as warnings for now. Here is how they are reported:
https://ci.adverity.net/blue/rest/organizations/jenkins/pipelines/PRESENSE-PR-FRONTEND-CHECKS/runs/1160/nodes/39/log/?start=0

And in IDE looks like this
image

**BUT:**
I am having an issue getting it perfectly right. 
Currently it makes a bold assumption that the codebase is in one folder down relative to the project root (for example `/src`) which might not be true for each project.
I need this check to cover some of the edge cases. 

It will work for presence, but I still need to test it with other projects.

---

@TkDodo commented on [Wed Mar 04 2020](https://github.com/pmedianetwork/adverity-presense-frontend/issues/327#issuecomment-594468188)

🎉 

---

@adamgajzlerowicz commented on [Thu Mar 05 2020](https://github.com/pmedianetwork/adverity-presense-frontend/issues/327#issuecomment-595094533)

I asked a question on eslint gitter channel. It is not possible to have rules in config. 

@TkDodo Suggested importing a rule locally, but that will work only locally [article](https://medium.com/neighborhoods-com-engineering-blog/custom-eslint-rules-for-faster-refactoring-2095e69bde08)

There is a way to do this in one repository. We need to change this repositories shape to become a plugin (instead of config)

This is explained [here](https://eslint.org/docs/developer-guide/working-with-plugins#configs-in-plugins)

### PR DESCRIPTION
**Pull request by adamgajzlerowicz**
Closes https://github.com/pmedianetwork/eslint-config/issues/65

Change structure of the project to become a plugin. 
Add custom rules

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,12 +1,15 @@
...
   "description": "Shared ESLint config for pmedianetwork projects.",
+  "author": "pmedianetwork",
   "main": "index.js",
   "scripts": {
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "test": "mocha ./tests/**/*.js -w"
   },
   "dependencies": {
+    "eslint-module-utils": "^2.5.2",
+    "requireindex": "~1.1.0",
```
Comment (TkDodo) [inline package.json]: strict dependencies please

File: tests/index.test.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+const { configs, rules } = require('../index');
+const { checkExists } = require('../utils');
```
Comment (TkDodo) [inline tests/index.test.js]: can we use es6 imports here?

File: package.json [hunk 3]
```diff
@@ -23,14 +26,8 @@
...
+    "mocha": "7.1.0"
   },
-  "author": "",
-  "license": "MIT",
-  "bugs": {
-    "url": "https://github.com/pmedianetwork/eslint-config/issues"
-  },
-  "homepage": "https://github.com/pmedianetwork/eslint-config#readme"
+  "license": "ISC"
 }
```
Comment (TkDodo) [inline package.json:49]: can we add prettier and linting to this repo now that we are actually writing code here? Not sure if we could lint it with "ourselves" .... linting the eslint repo _with_ the eslint repo is kinda meta 😅

**Review by TkDodo — CHANGES_REQUESTED** [review]

File: package.json (hunk 3 above)
Comment (adamgajzlerowicz) [reply package.json:49]: We can - I was thinking about it, but I wasn't sure if we event are fine with this massive refactor 😄 
I'd probably do it later in future as a separate issue, but that would be once we write more code here. What do you think?

File: tests/index.test.js (hunk 2 above)
Comment (adamgajzlerowicz) [reply tests/index.test.js]: I tried - doesn't work :(

File: package.json (hunk 1 above)
Comment (adamgajzlerowicz) [reply package.json]: 👍 
https://github.com/pmedianetwork/eslint-config/pull/64/commits/9a2edd9298020165d5aba95c994d9f3a1a91c610

File: tests/index.test.js (hunk 2 above)
Comment (adamgajzlerowicz) [reply tests/index.test.js]: I didn't want to spend to much time on the tooling for now.

File: package.json [hunk 4]
```diff
@@ -23,14 +26,8 @@
     "eslint-plugin-react": "7.16.0",
     "eslint-plugin-react-hooks": "2.2.0"
   },
-  "repository": {
-    "type": "git",
-    "url": "git+https://github.com/pmedianetwork/eslint-config.git"
+  "devDependencies": {
+    "mocha": "7.1.0"
```
Comment (TkDodo) [inline package.json]: can we use `jest` for testing like in all other repos?

**Review by TkDodo — CHANGES_REQUESTED** [review]

**Review by danielbartsch — APPROVED** [review]

File: package.json (hunk 4 above)
Comment (adamgajzlerowicz) [reply package.json]: The wrapper is made for mocha, but let me check if I can hack it

File: package.json (hunk 3 above)
Comment (TkDodo) [reply package.json:49]: okay, lets create a follow-up

File: tests/index.test.js (hunk 2 above)
Comment (TkDodo) [reply tests/index.test.js]: okok

**Comment by TkDodo** [comment]
> feature/version-3

just FYI, it's gonna be version 4, as the latest is now 3.2.0

File: .changeset/65-update-project-to-be-plugin.md [hunk 5]
```diff
@@ -0,0 +1,10 @@
+---
+issue: https://github.com/pmedianetwork/eslint-config/issues/65
+type: major
+---
+
+Changed structure of project from config, to a plugin.
```
Comment (TkDodo) [inline .changeset/65-update-project-to-be-plugin.md:6]: can we add a migration path here?

Comment (adamgajzlerowicz) [reply .changeset/65-update-project-to-be-plugin.md:6]: 👍 
https://github.com/pmedianetwork/eslint-config/pull/64/commits/928c075102879ea90f0b0c1b491aa7f0ba0a11c8

File: package.json (hunk 4 above)
Comment (adamgajzlerowicz) [reply package.json]: Documentation says `mocha` but replacing it for jest worked 😃 
https://github.com/pmedianetwork/eslint-config/pull/64/commits/cb8d029e9df365f3ca169cbe924dfa2567061a12

File: package.json (hunk 3 above)
Comment (adamgajzlerowicz) [reply package.json:49]: No need
https://github.com/pmedianetwork/eslint-config/pull/64/commits/9eca8320ec15989166ad516f413a5bc36980c0ca

Comment (TkDodo) [reply package.json:49]: wow, you are so fast! But I think we need the prettier-config that we have everywhere: single quotes as strings, no semicolons etc. We have this in `insights` in `package.json`:

```
  "prettier": {
    "printWidth": 120,
    "semi": false,
    "singleQuote": true,
    "trailingComma": "all",
    "tabWidth": 4
  },
```

File: package.json (hunk 4 above)
Comment (TkDodo) [reply package.json]: awesome!

File: README.md [hunk 6]
```diff
@@ -7,7 +7,7 @@ Use it by setting the extends field in your `.eslintrc`:
 
 ```js
 {
-  extends: '@pmedianetwork/eslint-config',
+  extends: 'pmedianetwork/recommended',
```
Comment (TkDodo) [inline README.md]: just a question: why did we get rid of the `@` prefix? Is that somehow required?

File: package.json [hunk 7]
```diff
@@ -1,12 +1,16 @@
 {
-  "name": "@pmedianetwork/eslint-config",
-  "version": "3.2.0",
+  "name": "eslint-plugin-pmedianetwork",
+  "version": "4.0.0",
```
Comment (TkDodo) [inline package.json]: no need to change the version manually - it will be done with the automated build process. If we have version 4 here now, version 5 will be released!

File: package.json [hunk 8]
```diff
@@ -17,20 +21,16 @@
...
     "eslint-plugin-prettier": "3.1.2",
     "eslint-plugin-react": "7.16.0",
     "eslint-plugin-react-hooks": "2.2.0"
   },
-  "repository": {
-    "type": "git",
-    "url": "git+https://github.com/pmedianetwork/eslint-config.git"
+  "devDependencies": {
+    "jest": "25.1.0",
+    "prettier": "^1.19.1"
```
Comment (TkDodo) [inline package.json]: another ^ snuck in. maybe we need an eslint rule for that, too 😄 

```suggestion
    "prettier": "1.19.1"
```

**Review by TkDodo — CHANGES_REQUESTED** [review]

File: README.md (hunk 6 above)
Comment (adamgajzlerowicz) [reply README.md]: Plugins for eslint normally use `eslint-plugin-xxx` naming convention. I just sticked with it. Do you think it is required to have it?

File: package.json (hunk 3 above)
Comment (adamgajzlerowicz) [reply package.json:49]: 👍 
https://github.com/pmedianetwork/eslint-config/pull/64/commits/a7455fda1e10bec289f02eb96f385204cd4813d7

File: package.json (hunk 7 above)
Comment (adamgajzlerowicz) [reply package.json]: https://github.com/pmedianetwork/eslint-config/pull/64/commits/9c05bc03230a0283e513fdd59ec9868051d9038e

Comment (adamgajzlerowicz) [reply package.json]: Ok - so they bump automatically - awesome!

File: README.md (hunk 6 above)
Comment (TkDodo) [reply README.md]: It would be nice to have all our own dependencies be in the same directory in `node_modules` (makes it easier to look it up, which we sometimes need to do):

Screenshot 2020-03-05 at 13 07 10

Comment (adamgajzlerowicz) [reply README.md]: Right - it took me couple of commits so I don't have one to paste, but it's sorted now :)

File: tests/index.test.js [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+    it('has configs', () => {
+        checkExists(configs)
+        checkExists(configs.recommended)
+        checkExists(configs.flow)
+        checkExists(configs.typescript)
+    })
+
+    it('has correct shapshot', () => {
+        expect(configs.recommended).toMatchSnapshot()
```
Comment (TkDodo) [inline tests/index.test.js:19]: will this not fail every time we change a recommended rule? What's the point?

**Review by TkDodo — CHANGES_REQUESTED** [review]

File: tests/index.test.js (hunk 9 above)
Comment (adamgajzlerowicz) [reply tests/index.test.js:19]: it is getting constructed - when we change something in the config we will know exactly what is changed. 

It is really simple to just pres `u`. This is an e2e check of the repo. I am also considering to add those tests for other rules.

**Review by TkDodo — APPROVED** [review]

File: tests/index.test.js [hunk 10]
```diff
@@ -15,7 +15,9 @@ describe('main export', () => {
         checkExists(configs.typescript)
     })
 
-    it('has correct shapshot', () => {
+    it('has correct shapshots', () => {
```
Comment (danielbartsch) [inline tests/index.test.js]:
```suggestion
    it('has correct snapshots', () => {
```

**Comment by danielbartsch** [comment]
can this be merged?

File: docs/rules/no-parent-imports.md [hunk 11]
```diff
@@ -0,0 +1,35 @@
+# no-parent-imports
+
+This will disallow imports that contain ../.. 
+
+### Rule Details
+
+When we want to go to far up to import a module, we should use an absolute import.
+
+### Examples of **incorrect** code for this rule:
```
Comment (twierzch) [inline docs/rules/no-parent-imports.md:9]: Here I think you meant  `correct` and down there the `incorrect` examples

File: lib/rules/no-absolute-internal-imports.js [hunk 12]
```diff
@@ -0,0 +1,57 @@
+const relative = require('path').relative
+
+const message = 'No internal modules imports with absolute path'
+
+const getWithoutLastSegment = path => {
+    const parts = path.split('/').reverse()
```
Comment (twierzch) [inline lib/rules/no-absolute-internal-imports.js:6]: Was this tested under windows? Because paths are not normalized so on windows they may have `\` for folders instead of `/` but I'm not 100% sure of that. I just remember that sometime ago I had some struggles with working with paths on both unix and windows systems

File: docs/rules/no-parent-imports.md (hunk 11 above)
Comment (adamgajzlerowicz) [reply docs/rules/no-parent-imports.md:9]: Thanks!
https://github.com/pmedianetwork/eslint-config/pull/64/commits/59553db45ded52fde5f77a5eacaf7998023fa335

File: lib/rules/no-absolute-internal-imports.js (hunk 12 above)
Comment (adamgajzlerowicz) [reply lib/rules/no-absolute-internal-imports.js:6]: Is this a requirement?

Comment (adamgajzlerowicz) [reply lib/rules/no-absolute-internal-imports.js:6]: https://adverity.slack.com/archives/CK95U7PGC/p1583746023093900?thread_ts=1583745883.093800&cid=CK95U7PGC

Comment (danielbartsch) [reply lib/rules/no-absolute-internal-imports.js:6]: It's not a requirement. As clarified in the slack devteam channel: https://adverity.slack.com/archives/CK95U7PGC/p1583745883093800

**Review by twierzch — APPROVED** [review]
